### ISSUE TITLE
powershell shebang ignored for modules

### ISSUE THREAD

**Issue by paylocity-sflanders** [ansible/ansible#50384]
##### SUMMARY
it appears the shebang in powershell based libraries is ignored and hardcoded in the python implementation.   I've been working with microsoft on figuring out some hanging issues with powershell core (https://github.com/PowerShell/PowerShell/issues/8545) and was wanting to use a workaround that appears to fix the issues but i'm unable to create one like:

````
#!powershell -c
````

When i look at the log i see the ssh call look like the following regardless of the shebang being there

````
 '/bin/sh -c '"'"'powershell /root/.ansible/tmp/ansible-tmp-1546104125.74-176453161072614/AnsiballZ_some_library_powershell.ps1 && sleep 0'"'"''
````

further investigation into the ansible code base showed this in module_common.py
````
    elif module_substyle == 'powershell':
        # Powershell/winrm don't actually make use of shebang so we can
        # safely set this here.  If we let the fallback code handle this
        # it can fail in the presence of the UTF8 BOM commonly added by
        # Windows text editors
        shebang = u'#!powershell'
````

as a test i updated that to

````
    elif module_substyle == 'powershell':
        # Powershell/winrm don't actually make use of shebang so we can
        # safely set this here.  If we let the fallback code handle this
        # it can fail in the presence of the UTF8 BOM commonly added by
        # Windows text editors
        shebang = u'#!powershell -c'
````

and my issues went away.   Is it possible to override this behavior some way?

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
executor/module_common.py

##### ANSIBLE VERSION

```paste below
ansible 2.7.2
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, Jul 13 2018, 13:06:57) [GCC 4.8.5 20150623 (Red Hat 4.8.5-28)]
```

##### CONFIGURATION

```paste below

```

##### OS / ENVIRONMENT
CentosOS 7

##### STEPS TO REPRODUCE
Simply run a basic powershell module against a linux box and turn on verbosity to see the ssh calls.  You'll see that the powershell call is hard coded to not allowing any shebang overrides.

##### EXPECTED RESULTS
i want the shebang to be honored

##### ACTUAL RESULTS
shebang is not honored

**Comment by ansibot** [comment]
Files identified in the description:
* [lib/ansible/executor/module_common.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/executor/module_common.py)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by jborean93** [comment]
@paylocity-sflanders sorry for not getting back to you earlier but my comments in https://github.com/ansible/ansible/issues/48881#issuecomment-440481672 still stand as current line for PowerShell Core on Linux. Ansible allows you to create a module in any language but support for using shared functions in module_utils are only for Python on the POSIX side and PowerShell for the Windows side. There are no plans to implement support for PowerShell on POSIX with PSCore for a few reasons;

* The majority of the PowerShell modules in module_utils are designed around Windows idioms and won't work on a POSIX host
* We don't even test/support PowerShell Core on Windows right now and I know that some aspects work but other areas do not
* The `Ansible.ModuleUtils.Legacy` module_util is an older way of managing modules, the replacement that has been recently added uses some Windows only functions to support some of the same features that are included in `basic.py` (Python module wrapper)
* There's nothing stopping a module util to add a Windows only feature breaking this scenario, we don't test this
* Even when support is added to configure the shebang this still won't work there are still other areas that will cause this to fail
* Currently PowerShell on Windows runs in a pipelined state whereas non-pipelining still needs to be supported for a POSIX host. This dramatically changes the behaviour around how a module is executed, your example has the non-pipelined scenario as the module is copied as a file to the host and executed there, whereas a pipelined scenario is run all in memory
* There's no technical reason why this can't be done but the current demand vs the cost of maintaining/testing this setup is just not there right now
* You can still run a module under PS Core on POSIX, just without the shared utils

We appreciate you working through this but unfortunately this isn't something we want to focus on right now.

However, we're absolutely always up for discussion.  Since this is a really busy project, we don't always see comments on closed tickets, but want to encourage
open dialog.  You can stop by the development list, and we'd be glad to talk about it - and we might even be persuaded otherwise!

   * https://groups.google.com/forum/#!forum/ansible-devel

Thank you once again for this and your interest in Ansible!

**Comment by paylocity-sflanders** [comment]
•	The majority of the PowerShell modules in module_utils are designed around Windows idioms and won't work on a POSIX host

I was under the impression these are helpers that are not required to use.  Is that not the case?  Obviously its a different case if the wrapper functionality is using windows-only functionality but from what i could tell it seemed to be trying to support really old PS versions.

•	We don't even test/support PowerShell Core on Windows right now and I know that some aspects work but other areas do not

The key with using Core at this point is sticking with the OS agnostic functionality for the most part because it doesn't have feature parity at this point.  We actually have ansible PS functionality that manage all our tower configuration using its services.  It currently gets executed on localhost because of the issue i hit.  Obviously OS specific stuff always has to be detected and coded differently

•	The Ansible.ModuleUtils.Legacy module_util is an older way of managing modules, the replacement that has been recently added uses some Windows only functions to support some of the same features that are included in basic.py (Python module wrapper)

Similar to my comment on these being helpers and not required.

•	There's nothing stopping a module util to add a Windows only feature breaking this scenario, we don't test this

I wouldn't expect it to detect this kind of dumbness.   If folks want to do bad things they will find a way.

•	Even when support is added to configure the shebang this still won't work there are still other areas that will cause this to fail

This is the one i'm most curious about.  What other issues?  Aside from the problem i saw with the hanging issue.   

•	Currently PowerShell on Windows runs in a pipelined state whereas non-pipelining still needs to be supported for a POSIX host. This dramatically changes the behaviour around how a module is executed, your example has the non-pipelined scenario as the module is copied as a file to the host and executed there, whereas a pipelined scenario is run all in memory

I assume you are referring to the fact that it bundles up everything in one file?  i was only aware of two execution modes in ansible.  one where it does the bundling and another where it just executes the script as-is and passes the json as a parameter on the powershell call.

•	There's no technical reason why this can't be done but the current demand vs the cost of maintaining/testing this setup is just not there right now

Its understandable.  We all have priorities.   The shebang thing seemed to be the main thing that was preventing our stuff from working without modifying that python file.

•	You can still run a module under PS Core on POSIX, just without the shared utils

We try to use the shared utils approach when possible since it makes it much easier to test and share to multiple libraries.

Thanks for the forum mention, i will post there as well.

**Comment by jborean93** [comment]
> I was under the impression these are helpers that are not required to use. Is that not the case? Obviously its a different case if the wrapper functionality is using windows-only functionality but from what i could tell it seemed to be trying to support really old PS versions.

They are not but having utils that support Windows and some that don't without a clearly defined line will confuse people. For example running a separate process in module is a common thing to do but the util that provides this functionality in a nice and easy wrapper https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/powershell/Ansible.ModuleUtils.CommandUtil.psm1 will only work on Windows. There's no technical reason why this shouldn't work except that we need to change the code so that is supports the POSIX spawn as well as Windows `CreateProcess`.

> The key with using Core at this point is sticking with the OS agnostic functionality for the most part because it doesn't have feature parity at this point

That is true but we don't even support PS Core on Windows right now in Ansible. Things "may" work but they can just as easily break in the future. The same problem exists here, we need to add a test matrix for this scenario before we can even consider it. The other thing we need to consider for PSCore is what does it offer over PS Desktop. The latter comes builtin to Windows and gives us most things we need to manage Windows hosts. While PS Core may introduce a killer feature we want there's nothing I know off right now that we need or hasn't already been solved by an existing Python module.

> Similar to my comment on these being helpers and not required.

I understand where you are coming from but it's not always black and white. Using the `Tmpdir` functionality added with the new PowerShell basic util we can see the code at https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/csharp/Ansible.Basic.cs#L119. This code will run when `Tmpdir` is accessed it will create a new tmpdir and set the file system permissions so only the current user has full access to the folder. This is done for security reasons and not just something we can ignore on the POSIX side. We would have to write the code to support this in .NET. This adds more complexity to an already complex bit of code and tests would be critical to ensure we actually test that code path.

> I wouldn't expect it to detect this kind of dumbness. If folks want to do bad things they will find a way.

I wouldn't expect it but unfortunately this isn't on everyone's mind when writing code :) Without tests to enforce this pattern there is not much we can do to guarantee it will never happen.

> This is the one i'm most curious about. What other issues? Aside from the problem i saw with the hanging issue.

One issue was that the `powershell.py` shell plugin didn't implement the `_SHELL_AND`, `_SHELL_OR`, etc class attributes that is called when Ansible is being executed. I came across another issue with powershell being hardcoded as the executable that is outside of `module_common` https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/shell/powershell.py#L65. Another issue is that `chmod` and `chown` is not implemented in `powershell.py` which we rely on to securely protect the temp dir that would store the module in a non-pipelined scenario. I haven't even tested some of the more complex scenarios like become and async so I have no idea if they work or not.

> I assume you are referring to the fact that it bundles up everything in one file? i was only aware of two execution modes in ansible. one where it does the bundling and another where it just executes the script as-is and passes the json as a parameter on the powershell call.

Sort off, historically Ansible copies the module (or wrapper) to the host and executes that (non-pipelined) but there exists another way of executing the module which we call pipelining. This is when the interpreter is called and the module payload is sent over the stdin pipe. This is a lot faster than the non-pipelined scenario as we don't need to copy a file, set permissions and then execute. Windows/powershell runs in a pipelined state by default, is disabled when `ANSIBLE_KEEP_REMOTE_FILES=1` is set, but it is the opposite in the POSIX side.

> Its understandable. We all have priorities. The shebang thing seemed to be the main thing that was preventing our stuff from working without modifying that python file.

I don't want to be the guy that says no but unfortunately in this case that will be the case :( As with most things when it comes to development it may be easy to implement one feature but the devil is in the details. Making sure we can test the scenario so it doesn't break in future changes and it doesn't add too much complexity is key. The PowerShell addition in Ansible is arguably still too complex and full of edge cases which we need to iron out but some of those edge cases are mostly due to how Windows differs from a POSIX environment.

> We try to use the shared utils approach when possible since it makes it much easier to test and share to multiple libraries.

That is definitely a good thing to do but right now Ansible doesn't offer a way to have shared libraries outside of Python (for POSIX) and PowerShell (for Windows). It would be great if we could implement something like this in a flexible manner but I'm just unsure as to how that would be possible as each language has their own import/sharing model making it hard to keep things generic. In this case, you miss out on some functions like `Exit-Json`, `Get-AnsibleParam` and so on but they are arguably not too complex anyway and can be replaced by a fairly small function.

**Comment by paylocity-sflanders** [comment]
Well i'll say your argument is much stronger than mine with these further details.  I'd have to agree i care more about reliability and compatibility of the base software so bringing support in for something new is always a lot of work.  It seemed much closer on the surface.  I do suspect the core line of products (.net, powershell, etc) will become the baseline at some point in the future.  Its definitely moving very quickly on the .net core side.

**Comment by jborean93** [comment]
Thanks for looking into it and posting the issue on PowerShell/PowerShell regarding the escape sequence. Good to know it's .NET causing the escape code sequence and not something in PowerShell.
